gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +273,439 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
-   *  both the function evaluation and its derivative(s) in X1 (and/or X2).
-   */
-  virtual Vector
-  evaluateError(const X1&, const X2&, boost::optional<Matrix&> H1 =
-      boost::none, boost::optional<Matrix&> H2 = boost::none) const = 0;
-
-private:
-
   /** Serialization function */
   friend class boost::serialization::access;
-  template<class ARCHIVE>
-  void serialize(ARCHIVE & ar, const unsigned int /*version*/) {
-    ar & boost::serialization::make_nvp("NoiseModelFactor",
-        boost::serialization::base_object<Base>(*this));
+  template <class ARCHIVE>
+  void serialize(ARCHIVE& ar, const unsigned int /*version*/) {
+    ar& boost::serialization::make_nvp(
+        "NoiseModelFactor", boost::serialization::base_object<Base>(*this));
   }
-}; // \class NoiseModelFactor2
+};  // \class NoiseModelFactorN
 
 /* ************************************************************************* */
-/** A convenient base class for creating your own NoiseModelFactor with 3
- * variables.  To derive from this class, implement evaluateError(). */
-template<class VALUE1, class VALUE2, class VALUE3>
-class NoiseModelFactor3: public NoiseModelFactor {
-
-public:
+/** @deprecated: use NoiseModelFactorN, replacing .key() with .key<1> and X1
+ * with X<1>.
+ * A convenient base class for creating your own NoiseModelFactor
+ * with 1 variable.  To derive from this class, implement evaluateError().
+ */
+template <class VALUE>
+class NoiseModelFactor1 : public NoiseModelFactorN<VALUE> {
+ public:
+  // aliases for value types pulled from keys
+  using X = VALUE;
+
+ protected:
+  using Base = NoiseModelFactor;  // grandparent, for backwards compatibility
+  using This = NoiseModelFactor1<VALUE>;
+
+ public:
+  // inherit NoiseModelFactorN's constructors
+  using NoiseModelFactorN<VALUE>::NoiseModelFactorN;
+  ~NoiseModelFactor1() override {}
 
-  // typedefs for value types pulled from keys
-  typedef VALUE1 X1;
-  typedef VALUE2 X2;
-  typedef VALUE3 X3;
+  /** method to retrieve key */
+  inline Key key() const { return this->keys_[0]; }
 
-protected:
-
-  typedef NoiseModelFactor Base;
-  typedef NoiseModelFactor3<VALUE1, VALUE2, VALUE3> This;
+ private:
+  /** Serialization function */
+  friend class boost::serialization::access;
+  template <class ARCHIVE>
+  void serialize(ARCHIVE& ar, const unsigned int /*version*/) {
+    ar& boost::serialization::make_nvp(
+        "NoiseModelFactor", boost::serialization::base_object<Base>(*this));
+  }
+};  // \class NoiseModelFactor1
 
-public:
+/* ************************************************************************* */
+/** @deprecated: use NoiseModelFactorN, replacing .key1() with .key<1> and X1
+ * with X<1>.
+ * A convenient base class for creating your own NoiseModelFactor
+ * with 2 variables.  To derive from this class, implement evaluateError().
+ */
+template <class VALUE1, class VALUE2>
+class NoiseModelFactor2 : public NoiseModelFactorN<VALUE1, VALUE2> {
```

Comment:
Add the `GTSAM_DEPRECATED` attribute to these classes?

```cpp
class GTSAM_DEPRECATED NoiseModelFactor2 ...
```

Reply:
done